bundle/endpoint/client.go

```diff
@@ -0,0 +1,117 @@
+// Package endpoint implements a SPIFFE bundle endpoint client.
+package endpoint
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/spiffe/go-spiffe/bundle"
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/zeebo/errs"
+)
+
+// SPIFFEAuthConfig contains the configuration of the SPIFFE authentication
+// mechanism for authenticating the bundle endpoint.
+type SPIFFEAuthConfig struct {
+	// EndpointSpiffeID is the expected SPIFFE ID of the endpoint server.
+	EndpointSpiffeID string
+
+	// RootCAs is the set of root CA certificates used to authenticate the
+	// endpoint server.
+	RootCAs []*x509.Certificate
+}
+
+// ClientConfig contains the configuration to create a bundle endpoint client.
+type ClientConfig struct {
+	// EndpointAddress is the bundle endpoint for the trust domain.
+	EndpointAddress string
+
+	// SPIFFEAuth contains required configuration to authenticate the endpoint
+	// using SPIFFE authentication. If unset, it is assumed that the endpoint
+	// is authenticated via Web PKI.
+	SPIFFEAuth *SPIFFEAuthConfig
+}
+
+// Client is used to fetch a bundle and metadata from a bundle endpoint
+type Client interface {
+	FetchBundle(context.Context) (*bundle.Bundle, error)
+}
+
+type client struct {
+	c      ClientConfig
+	client *http.Client
+}
+
+// NewClient creates a new bundle endpoint client
+func NewClient(config ClientConfig) (Client, error) {
+	if config.EndpointAddress == "" {
+		return nil, errs.New("bundle endpoint address is required")
+	}
+
+	httpClient := &http.Client{}
+	if config.SPIFFEAuth != nil {
+		spiffeID := config.SPIFFEAuth.EndpointSpiffeID
+		if spiffeID == "" {
+			return nil, errs.New("bundle endpoint spiffe ID is required")
+		}
+		if len(config.SPIFFEAuth.RootCAs) == 0 {
+			return nil, errs.New("an initial up-to-date bundle from the remote trust domain is required")
+		}
+
+		peer := &spiffe_tls.TLSPeer{
```

Comment:
We should avoid go-spiffe/tls since it is deprecated and does not do as much validation as it should. We should probably use an adapter over `spiffe.VerifyPeerCertificate`.  I'm ok with this going in as-is but we'll want to look at replacing this part soon.